processor/metricstransformprocessor/metrics_transform_processor_helper.go

```diff
@@ -0,0 +1,375 @@
+	labelIdxs, labels := mtp.getLabelIdxs(metric, labelSet)
+	// key is a composite of the label values as a single string
+	// keyToTimeseriesMap groups timeseries by the label values
+	// keyToLabelValuesMap maps the keys to the actual label values objects
+	keyToTimeseriesMap, keyToLabelValuesMap := mtp.groupTimeseries(metric, labelIdxs, op.NewValue, op.AggregatedValues)
+	// merge groups of timeseries
+	newTimeseries := mtp.aggregateTimeseriesGroups(keyToTimeseriesMap, keyToLabelValuesMap, op.AggregationType, metric.MetricDescriptor.Type)
+	metric.Timeseries = newTimeseries
+	if op.Action == AggregateLabels {
+		metric.MetricDescriptor.LabelKeys = labels
+	}
+}
+
+// createLabelValueMapping creates a label value mapping from old value to new value based on valueActions
+// Returns the mapping
+func (mtp *metricsTransformProcessor) createLabelValueMapping(valueActions []ValueAction) map[string]string {
+	mapping := make(map[string]string)
+	for _, valueAction := range valueActions {
+		mapping[valueAction.Value] = valueAction.NewValue
+	}
+	return mapping
+}
+
+// sliceToSet converts slice of strings to set of strings
+// Returns the set of strings
+func (mtp *metricsTransformProcessor) sliceToSet(slice []string) map[string]bool {
+	set := make(map[string]bool, len(slice))
+	for _, label := range slice {
+		set[label] = true
+	}
+	return set
+}
+
+// getLabelIdxs gets the indices of the labelSet labels' indices in the metric's descriptor's labels field
+// Returns the indices slice and a slice of the actual labels selected by this slice of indices
+func (mtp *metricsTransformProcessor) getLabelIdxs(metric *metricspb.Metric, labelSet map[string]bool) ([]int, []*metricspb.LabelKey) {
+	labelIdxs := make([]int, 0)
+	labels := make([]*metricspb.LabelKey, 0)
+	for idx, label := range metric.MetricDescriptor.LabelKeys {
+		_, ok := labelSet[label.Key]
+		if ok {
+			labelIdxs = append(labelIdxs, idx)
+			labels = append(labels, label)
+		}
+	}
+	return labelIdxs, labels
+}
+
+// groupTimeseries groups all timeseries in the metric that will be aggregated together based on the selected labels' values indicated by labelIdxs OR on the entire label values after replacing the aggregatedValues by newValue.
+// Depending on if newValue is set, the approach to group timeseries is different.
+// Returns a map from keys to groups of timeseries and a map from keys to the corresponding label values
+func (mtp *metricsTransformProcessor) groupTimeseries(metric *metricspb.Metric, labelIdxs []int, newValue string, aggregatedValues []string) (map[string][]*metricspb.TimeSeries, map[string][]*metricspb.LabelValue) {
+	// key is a composite of the label values as a single string
+	// keyToTimeseriesMap groups timeseries by the label values
+	keyToTimeseriesMap := make(map[string][]*metricspb.TimeSeries)
+	// keyToLabelValuesMap maps the keys to the actual label values objects
+	keyToLabelValuesMap := make(map[string][]*metricspb.LabelValue)
```

Comment:
Why not just have one map that stores both of these?

Reply:
You mean a key mapping to two lists, one being the timeseries and labelValues? This is definitely a valid solution, but does that compromise the readability of this data structure? What do you think?

Reply:
I think it makes it a lot easier to read 😄

You'd be storing all related data as one parameter instead of passing around two maps that need to be kept together.

Reply:
I guess I am speaking from the perspective of the data structure complexity. Since the two slices are of different types, then merging these into one map would result in a map that uses an interface type slice, which looks like `map[string][]interface{}{[]*metricspb.TimeSeries, []*metricspb.LabelValue}`, and when I want to get these, I would do mapping[0] and mapping[1]. I might also be thinking about it in a complicated way though.

Reply:
Use an existing type